quodlibet/quodlibet/ext/events/auto_update_tags_in_files.py

```diff
@@ -0,0 +1,246 @@
+    PLUGIN_DESC = _("When songs were played, update the tags in their files. "
+                    "This will ensure play counts and ratings are up to date.")
+    PLUGIN_ICON = Icons.DOCUMENT_SAVE
+
+    def PluginPreferences(self, _):
+        return AutoUpdateTagsPrefs()
+
+    def plugin_on_song_ended(self, song, skipped):
+        if song is None or not is_writable(song):
+            return
+
+        strategy = CONFIG.update_strategy
+        if strategy == UpdateStrategy.AFTER_PLAY_NOT_SKIP:
+            if not skipped:
+                self._try_to_update_song(song)
+            return
+        elif strategy == UpdateStrategy.AFTER_PLAY_OR_SKIP:
+            self._try_to_update_song(song)
+            return
+
+        self._update_album_if_fully_rated(song.album_key)
+
+    def _try_to_update_song(self, song_wrapper):
+        try:
+            song_wrapper._needs_write = True
+            self._write_tags_to_files([song_wrapper])
+        except Exception as e:
+            ErrorMessage(app.window,
+                         SONG_ERROR_FMT.format(song_wrapper, self.PLUGIN_NAME),
+                         "{}\n\n{}".format(e, format_exc())).run()
+
+    def _update_album_if_fully_rated(self, album_key):
+        album = app.library.albums.get(album_key, None)
+        if album is None:
+            return
+
+        songs = album.songs
+        # first check for ratings to avoid costly file checks
+        if not songs or not all(song.has_rating for song in songs):
+            return
+
+        one_or_more_need_update = False
+        for song in songs:
+            if one_or_more_need_update:
+                continue  # avoid files access
+            elif not is_rating_and_play_count_in_file(song['~filename']):
+                one_or_more_need_update = True
+
+        if not one_or_more_need_update:
+            return
+
+        # at least one song has no ratings or play counts stored in files
+        try:
+            self._update_album(songs)
+        except Exception as e:
+            ErrorMessage(app.window,
+                         ALBUM_ERROR_FMT.format(album_key, self.PLUGIN_NAME),
+                         "{}\n\n{}".format(e, format_exc())).run()
+
+    def _update_album(self, songs):
+        song_wrappers = []
+
+        req_play_counts_above_zero = CONFIG.ensure_play_counts_above_zero
+        for song in songs:
+            if req_play_counts_above_zero and not song.get("~#playcount", 0):
+                song['~#playcount'] = 1
+
+            wrapper = SongWrapper(song)
+            wrapper._needs_write = True
+            song_wrappers.append(wrapper)
+
+        self._write_tags_to_files(song_wrappers)
+
+    def _write_tags_to_files(self, song_wrappers):
+        background_check_wrapper_changed(app.library, song_wrappers)
+
+
+def is_rating_and_play_count_in_file(path):
+    save_email = config.get("editing", "save_email").strip()
+
+    f = mutagen.File(path)
+
+    if isinstance(f, MP3):
```

Comment:
AFAICS `AudioFile` contains no way to test whether certain tags are written to the file. (i.e. it can check whether the song has a `~#rating`, but not whether that is written to the file)

At first I thought there was some way via `reload` but besides being a bit hacky it seems far worse performance wise, since reading, handling and replacing all of the data from the file is unnecessary for the small check we need to make.


Interestingly, `reload` always replaces the values of `~#rating` and `~#playcount`, that were loaded from the file, with the stored one (which may not ever have been written to the file, e.g. if a user just rated), since they're in the `MIGRATE` set.

Reply:
>  I guess we could add has_rating_and_playcount_in_file method to AudioFile and then overwrite it in ID3File and MutagenVCFile (the only ones that support storing ratings it seems)

Sounds like a good idea, but I guess probably a separate issue at this point...